client/client-lib/src/mint/backup.rs

```diff
@@ -216,5 +330,393 @@ impl EcashBackup {
     }
 }
 
+/// The state machine used for fast-fowarding backup from point when it was taken to the present time.
+struct EcashRecoveryNonceTracker {
+    /// Nonces that we track that are currently spendable.
+    spendable_note_by_nonce: HashMap<Nonce, (Amount, SpendableNote)>,
+
+    /// Outputs (by `OutPoint`) we track federation member confirmations for blind nonces.
+    ///
+    /// Once we get enough confirmation (valid shares), these become new spendable notes.
+    ///
+    /// Note that `NoteIssuanceRequest` is optional, as sometimes we might need
+    /// to handle a tx where only some of the blind nonces were in the pool.
+    /// A `None` means tha this blind nonce/message is there only for validation
+    /// purposes, and will actually not create a `spendable_note_by_nonce`
+    #[allow(clippy::type_complexity)]
+    pending_outputs: HashMap<
+        OutPoint,
+        (
+            TieredMulti<(BlindedMessage, Option<NoteIssuanceRequest>)>,
+            HashMap<PeerId, Vec<tbs::BlindedSignatureShare>>,
+        ),
+    >,
+
+    /// Next nonces that we expect might soon get used.
+    /// Once we see them, we move the tracking to `pending_outputs`
+    ///
+    /// Note: since looking up nonces is going to be the most common operation
+    /// the pool is kept shared (so only one lookup is enough), and replenishment
+    /// is done each time a note is consumed.
+    pending_nonces: HashMap<BlindedMessage, (NoteIssuanceRequest, NoteIndex, Amount)>,
+
+    /// Tail of `pending`. `pending_notes` is filled by generating note with this index
+    /// and incrementing it.
+    next_pending_note_idx: Tiered<NoteIndex>,
+
+    /// `LastECashNoteIndex` but tracked in flight. Basically max index of any note that got
+    /// a partial sig from the federation (initialled from the backup value).
+    /// TODO: One could imagine a case where the note was issued but not get any partial sigs yet.
+    /// Very unlikely in real life scenario, but worth considering.
+    last_mined_nonce_idx: Tiered<NoteIndex>,
+
+    /// Threshold
+    threshold: usize,
+
+    /// The **mint** (not root) derived secret used to derive notes
+    secret: DerivableSecret,
+
+    /// Public key shares for each peer
+    ///
+    /// Used to validate contributed consensus items
+    pub_key_shares: BTreeMap<PeerId, Tiered<PublicKeyShare>>,
+
+    /// Aggregate public key for each amount tier
+    tbs_pks: Tiered<AggregatePublicKey>,
+
+    /// The number of nonces we look-ahead when looking for mints (per each amount).
+    gap_limit: usize,
+}
```

Comment:
I wonder to which degree we should save partial progress to disk in case the recovery process gets killed by  the OS and later restarted.

Reply:
As this struct is side-effect free pure state machine, if we add `Encodable` the user of it can write it periodically to storage along with the last epoch it processed and just continue, so it should be easy to add.

We can decide after we test how long it takes. Probably we will need it, as phones have a tendency to kill apps on demand.